File: LeetCode/20.js
// Valid Parentheses
const isValid = function (s) {
	const openers = new Set(["{", "[", "("]);
	const closers = new Set(["}", "]", ")"]);
	const map = new Map();

	map.set("{", "}");
	map.set("(", ")");
	map.set("[", "]");

	const stack = [];

	for (const bracket of s) {
		if (openers.has(bracket)) {
			stack.push(bracket);
		} else if (closers.has(bracket)) {
			const fromStack = stack.pop();
			if (map.get(fromStack) !== bracket) {
				return false;
			}
		}
	}
	return stack.length === 0;
};

File: LeetCode/src/20.cpp
// Valid Parentheses
#include <iostream>
#include <stack>
using namespace std;

class Solution {
public:
  bool isValid(string s) {
    if (s.size() % 2 != 0) {
      return false;
    }
    stack<char> stack;
    for (char c : s) {
      if (c == '(' || c == '{' || c == '[') {
        stack.push(c);
      } else {
        if (stack.empty()) {
          return false;
        }
        if (c == ')' && stack.top() == '(') {
          stack.pop();
        } else if (c == '}' && stack.top() == '{') {
          stack.pop();
        } else if (c == ']' && stack.top() == '[') {
          stack.pop();
        } else {
          return false;
        }
      }
    }
    return stack.empty();
  }
};
